pkg/prowloader/prow.go

```diff
@@ -197,18 +202,42 @@ func (pl *ProwLoader) syncPRStatus() error {
 	}
 
 	for _, pr := range pulls {
+
+		logger := log.WithField("org", pr.Org).
+			WithField("repo", pr.Repo).
+			WithField("number", pr.Number).
+			WithField("sha", pr.SHA)
+
 		mergedAt, err := pl.githubClient.GetPRSHAMerged(pr.Org, pr.Repo, pr.Number, pr.SHA)
 		if err != nil {
-			log.WithError(err).Warningf("could not fetch pull request status from GitHub; org=%q repo=%q number=%q sha=%q", pr.Org, pr.Repo, pr.Number, pr.SHA)
+			logger.WithError(err).Warningf("could not fetch pull request status from GitHub; org=%q repo=%q number=%q sha=%q", pr.Org, pr.Repo, pr.Number, pr.SHA)
+			// return or keep trying next record?
 			return err
```

Comment:
Here, I was curious if we should really exit or just continue on to the next pull

Reply:
https://docs.github.com/en/rest/rate-limit?apiVersion=2022-11-28

Let me see about adding something into our githubClient to check

Reply:
Added in some rate limit checking and ran syncPRStatus locally with a check on the remaining calls every 10 seconds to document that we are churning through them.  I think I have another approach that should help.

`INFO[2023-02-06T13:26:49.896-05:00] Unmerged pull request count: 14962           
INFO[2023-02-06T13:27:00.045-05:00] Github Limit:5000, Remaining:4966            
INFO[2023-02-06T13:27:10.219-05:00] Github Limit:5000, Remaining:4934            
INFO[2023-02-06T13:27:20.592-05:00] Github Limit:5000, Remaining:4901            
INFO[2023-02-06T13:27:30.717-05:00] Github Limit:5000, Remaining:4869   `

Reply:
Does the rate limiter get used on syncPRStatus? I only see it used in the commenter code.

Of course we can deal with it separately from this PR. I ran the PR commenter locally and I think it's pretty good and we could address anything else in follow-ups.

Reply:
I have some refactoring that should help.  We can get the closed PRs sorted based on most recent updates and get the last day or twos worth, cache them and then check the cache for each unmerged PR/SHA combo we have in our prow_pull_requests table.  Anything not closed / merged & updated recently we will just skip over.

Part of why there are so many is we aren't checking to see if the PR has merged but if a specific SHA for that PR has merged https://github.com/openshift/sippy/blob/a6da98f18b43c65a7e83fb722ea8976aa5079bb6/pkg/prowloader/prow.go#L210

So the 'unmerged' results may remain high for PRs that have multiple commits in them prior to merge or just get closed but the caching of closed PRs with recent updates should minimize the calls we have to make processing the results.


INFO[2023-02-06T16:55:51.83-05:00] Unmerged pull request count: 14885           
INFO[2023-02-06T16:56:02.278-05:00] Processing idx: 4                            
INFO[2023-02-06T16:56:02.347-05:00] Github Limit:5000, Remaining:4980            
INFO[2023-02-06T16:56:12.376-05:00] Processing idx: 10                           
INFO[2023-02-06T16:56:12.435-05:00] Github Limit:5000, Remaining:4975            
INFO[2023-02-06T16:56:23.576-05:00] Processing idx: 25                           
INFO[2023-02-06T16:56:23.666-05:00] Github Limit:5000, Remaining:4969            
INFO[2023-02-06T16:56:34.607-05:00] Processing idx: 39                           
INFO[2023-02-06T16:56:34.682-05:00] Github Limit:5000, Remaining:4962  

...

INFO[2023-02-06T17:07:49.667-05:00] Processing idx: 14203                        
INFO[2023-02-06T17:07:49.751-05:00] Github Limit:5000, Remaining:3871            
INFO[2023-02-06T17:08:00.235-05:00] Processing idx: 14272                        
INFO[2023-02-06T17:08:00.312-05:00] Github Limit:5000, Remaining:3855            
INFO[2023-02-06T17:08:11.144-05:00] Processing idx: 14436                        
INFO[2023-02-06T17:08:11.208-05:00] Github Limit:5000, Remaining:3831            
INFO[2023-02-06T17:08:21.392-05:00] Processing idx: 14490                        
INFO[2023-02-06T17:08:21.469-05:00] Github Limit:5000, Remaining:3820            
INFO[2023-02-06T17:08:31.714-05:00] Processing idx: 14701                        
INFO[2023-02-06T17:08:31.777-05:00] Github Limit:5000, Remaining:3801  

INFO[2023-02-06T17:09:22.333-05:00] Unmerged pull request count: 14809

Reply:
We can actually do better here.  If we know what has recently merged we can get the mergedAt as well as the mergedSHA.  Then just compare that SHA with our record, if it matches update the record, otherwise we know the SHA in our record is not the one that merged.

Down to under 200 hits processing 14k records.

INFO[2023-02-06T18:15:12.157-05:00] Processing idx: 4                            
INFO[2023-02-06T18:15:12.225-05:00] Github Limit:5000, Remaining:4625            
INFO[2023-02-06T18:15:23.179-05:00] Processing idx: 11                           
INFO[2023-02-06T18:15:23.241-05:00] Github Limit:5000, Remaining:4620            
INFO[2023-02-06T18:15:33.751-05:00] Processing idx: 26                           
INFO[2023-02-06T18:15:33.817-05:00] Github Limit:5000, Remaining:4615   

...

INFO[2023-02-06T18:21:43.575-05:00] Processing idx: 11642                        
INFO[2023-02-06T18:21:43.663-05:00] Github Limit:5000, Remaining:4449            
INFO[2023-02-06T18:21:53.964-05:00] Processing idx: 12373                        
INFO[2023-02-06T18:21:54.026-05:00] Github Limit:5000, Remaining:4444            
INFO[2023-02-06T18:22:05.087-05:00] Processing idx: 14225                        
INFO[2023-02-06T18:22:05.155-05:00] Github Limit:5000, Remaining:4438

Reply:
That's awesome!